tools/iree-e2e-matmul-test.c

```diff
@@ -276,6 +280,21 @@ static iree_status_t get_matmul_sizes(
   return iree_ok_status();
 }
 
+// Helper for reference_matmul_element. f16 case.
+static void reference_matmul_element_f16(
+    iree_hal_dim_t m_size, iree_hal_dim_t k_size, iree_hal_dim_t n_size,
+    iree_hal_element_type_t lhs_type, iree_hal_element_type_t rhs_type,
+    uint16_t* lhs_data, uint16_t* rhs_data, uint16_t* acc_data,
+    uint16_t* result_data, iree_hal_dim_t m, iree_hal_dim_t n) {
+  float acc = iree_math_f16_to_f32(acc_data[n + m * n_size]);
+  for (iree_hal_dim_t k = 0; k < k_size; ++k) {
+    float lhs_value = iree_math_f16_to_f32(lhs_data[k + m * k_size]);
+    float rhs_value = iree_math_f16_to_f32(rhs_data[n + k * n_size]);
+    acc += lhs_value * rhs_value;
```

Comment:
This reference matmul implementation is really a `f32 += f16 * f16` matmul, not a `f16 += f16 * f16`, since `acc` has type `float`. I believe that in order to make it exactly `f16 += f16 * f16`, you would have to do something like this: first have a helper function to round a float to the nearest f16,

float round_to_nearest_f16_representable(float x) {
  return iree_math_f16_to_f32(iree_math_f32_to_f16(x));
}

then insert that around both the multiplication and the addition in `acc += lhs_value * rhs_value`.... is that correct?

Side note: the reason why this isn't caught by the test despite the test requiring bit-for-bit exactness is that the test input values are special enough that all intermediate values here are exactly f16-representable anyway. I don't think this is sufficient reason to reconsider that, just worth noting.

Reply:
> Do we have similar implementations of [half](https://github.com/NVIDIA/cutlass/blob/master/include/cutlass/half.h) and [bfloat16](https://github.com/NVIDIA/cutlass/blob/master/include/cutlass/bfloat16.h) types that we can use in iree?

Good question, i don't know! @benvanik , @antiagainst might.

> It could help handle these concerns by overloading multiply and add operators for the respective data types. For example [half_t operator*(half_t const& lhs, half_t const& rhs)](https://github.com/NVIDIA/cutlass/blob/master/include/cutlass/half.h#L797)

Noooooooooooo :-D explicitness is good here. The overriding concern here is to keep this code easy to read in the sense of proving that it's correct, not just easy to glance over casually --- operator overloading would help the latter, not the former.

> Yes, I agree this worth noting and there are trade-offs of testing f16 arithmetic using tolerance on real inputs vs. bit-exactness on small integers.

So I was serious in the last part of my comment above,

> > Or OK, we can plausibly argue here that we don't need to worry about any of this at all, because this reference implementation is meant to be used only in the context of such special test values, but then this deserves at least a comment.

That is, seriously, it's OK to explicitly make this reference matmul code specific to the special flavor of test values that we use here, and take advantage of that to not need f16 arithmetic at all in the reference code. I'm OK either way, up to you.

Reply:
Just had a conversation with @manishucsd . Here's the summary of the thinking we've arrived at.

We need to keep consistency between the test code, including its reference matmul implementation, and the test input values.

Choose one route between A and B:

A. If we have integral test input values allowing us to require bit-exactness, then it makes sense to have only one reference matmul implementation, multiplying integer matrices. We still nead to load matrix data from the test matrices in various element types, but we can cast that to int and feed that to a single shared reference matmul.

B. If we have more general test input values causing floating-point rounding to occur, then more factors come into the picture. We were so far papering over them by using a fuzzy comparison, but this PR switches (at least for f16, but I do believe this is a good change to make for all types) to exact comparisons, so we are not papering over anything anymore and we need reference code to be bit-exact with the actual device code. Now, besides the data type (f16 vs f32) device code will also differ in at least fused-vs-unfused multiply-add, so we would need at least 2x more reference code paths for that. And that's not even starting to look into details of device-specific rounding modes!

The point is that if we keep separate reference code paths as in B. then we need test coverage ensuring that the distinctions justifying these separate code paths, are actually correctly handled, in particular we need more than integral test values. This circles back to the specific issue noted at the start of this thread, the f16 reference code was actually accumulating in f32 and the test didn't notice due to the integral test values.

Reply:
Thanks @bjacob for summarizing this. I have made the change to fill the buffers for floating point inputs with small integers, create a reference matmul which takes buffers as void*, and casts the values to int32_t. There are various cast on reads.

Reply:
On C++ struct for half_t datatype and explicitness of conversion. Please review the pseudo code below:

Given a `half_t` implementation with conversions from `F32->F16` and `F16->F32`. Also operator for matrix-multiply as below:
```h 
/// half.h
struct half_t {
/// half_t bits stored in 16-bit uint16_t
uint16_t storage;

/// Explicit conversion ctor
explicit half_t(float x);

/// F32->F16 conversion
operator float() const;
};

/// Multiplication on half_f numbers.
half_t operator*(half_t const& lhs, half_t const& rhs) {
#if HW_FP16_MUL_SUPPORTED
   __hw_mul_on_uint16_(lhs.storage, rhs.storage);
#else
  return half_t(float(lhs) * float(rhs));
#endif
}

// Output stream operator 
inline
std::ostream & operator<<(std::ostream &out, half_t const &x) {
  return out << float(x);
}


The application/user code using the above half_t type:
```c++
#include <half_h>
...
half_t opA(1.0f);
half_t opB(2.0f);
half_t opC;

// Multiplication on half_t datatypes
opC = opA * opB;

// Print half results
std::cout << opC
```
I think using structured `half_t` type will make handling F16 datatype in the application more graceful and the operations on a typed c++ langugage will make sure the programmer know that multiply is happening on half_t types.